### PR TITLE
Testwhatv2

### PR DESCRIPTION
**Pull request by filipsch**

### CODE REVIEW

**Comment by filipsch** [comment]
@vincentvankrunkelsven any feedback is still welcome, sorry I already merged!

File: R/messages.R [hunk 1]
```diff
@@ -1,84 +1,245 @@
...
+    return(capitalize(trim(details)))
   }
-  return(msg)
   
+  total_msg <- ""
+  for (det in details) {
+    if (!is.null(det$message)) {
+      msg <- det$message
+    } else {
+      msg <- ""
```
Comment (vincentvankrunkelsven) [inline R/messages.R]: Couldn't you avoid this giant if statement by using an S3 function:
```
class(det) <- det$type
msg <- generate_message(det)
```
You could use S3 functions to split further using `det$case` as well, or just use `switch(...)` statements there?

File: R/check-control.R [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+                            message = not_found_msg,
+                            append = append,
+                            pd = NULL)
+
+  extract_fun <- switch(type,
+                        `if` = extract_if,
+                        `for` = extract_for,
+                        `while` = extract_while)
+  student_structs <- extract_fun(student_pd)
+  solution_structs <- extract_fun(solution_pd)
```
Comment (vincentvankrunkelsven) [inline R/check-control.R]: I'm wondering whether you could make `extract()` into a S3 function which takes a state as a first argument and enriches this with extracted statements. 
E.g.: 
```
extract.ControlState <- function(state, type, ...) {
  statement_key <- paste0(type,"_statements")
  control_statements <- state$get(statement_key)
  if (is.null(control_statement)) {
    return(control_statements)
  } else {
    ... # Extraction logic depending on type
    state$set(statement_key = control_statements)
    return(control_statements)
  }
}
```
Then you could add an extractor per State, seems semantically logical (now there's a `find_function_calls` somewhere, which is actually the function call extractor I think?). Also it's reusable if you want to check multiple things on the same state.

File: R/state.R [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+#' Override the solution code and pd in a state
+#' 
+#' Produces a new state with a custom solution code. Mostly useful inside
+#' \code{test_or}, if you want to test for different cases.
+#' 
+#' @param state the state to create a substate from
+#' @param code the solution code to put into the state
+#'   
+#' @export
+override_solution <- function(state, code) {
```
Comment (vincentvankrunkelsven) [inline R/state.R]: Can we override (or rather extend) `solution_env` as well?

File: R/check-call.R [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+  
+  res <- logical(length(student_calls))
+  details <- NULL
+  for (i in seq_along(student_calls)) {
+    student_call <- student_calls[[i]]
+    if (isTRUE(is.na(student_call))) next
+    
+    # If no hits, use details of the first try
+    if (is.null(details)) {
+      arg_state$set_details(pd = student_call$pd)
```
Comment (vincentvankrunkelsven) [inline R/check-call.R]: Is this the complete parsed data, or part of it?

**Comment by vincentvankrunkelsven** [comment]
I went through the code (skimmed through it, was too much to look into detail). I really like the new system. Sorry I didn't give more feedback (and sooner), don't really feel 100% qualified to give a lot of feedback on R code :-).

File: R/state.R (hunk 3 above)
Comment (filipsch) [inline R/state.R]: This is not yet possible, but a good idea; I'll add it!

File: R/check-call.R (hunk 4 above)
Comment (filipsch) [inline R/check-call.R]: Only the parse data of the student_call

File: R/messages.R (hunk 1 above)
Comment (filipsch) [inline R/messages.R]: Truth, I went for quick n dirty here. I'll look into it.

**Pull request merged**